### ISSUE TITLE
Add unit tests for PGD implementation in PyTorch

### ISSUE THREAD

**Issue by AlexeyKurakin** [tensorflow/cleverhans#1040]

**Comment by michaelshiyu** [comment]
@AlexeyKurakin Can I work on this? Will copying the corresponding test suite for the current TF-based PGD do for now? Thanks.

**Comment by AlexeyKurakin** [comment]
@michaelshiyu  you're welcome to work on this.
I would advise against reusing tests for TF-Based PGD. The main problem with current tests is that they take too long to run and we generally want to reduce test time. Most of the time during tests is taken for training of the models.

There are few ideas I have regarding tests:
* Provide some small pre-trained model as a test data.
* Make up some artificial model (like linear model) for which you can make adversarial examples analytically and compare that PGD finds the same adversarial examples
* Make tests only for subroutines which are easy to test

**Comment by michaelshiyu** [comment]
Thanks for the details. 

So to clarify, the main issue with the current test suite is the time needed for running the tests? But the coverage and the test methods are okay in general?

Thanks.

**Comment by AlexeyKurakin** [comment]
Yes, the current test coverage is fine. Main issue is that they are too slow.

Regarding test methods. Some adaptation of test code will be needed due to the changes in API, for examples: model is now callable which takes input and outputs logits, there are no "generate" and "generate_np" just simple method to run the attack.

**Comment by michaelshiyu** [comment]
Great! I’ll get the basic stuff done first and tag you for a review. Thanks a lot!
